### PR TITLE
ENH: Sync features for lxt_ttc

### LINKED ISSUE #770: Issues with lxt, txt, lxt_ttc in XPP and XCS
**Bug Report**

Some sequences of move, set_current_position, move results in strange and unpredictable behavior for lxt_ttc.

Example issue paraphrased from meeting notes:
```
set lxt to zero
set txt to zero
move ttc to 10ps
txt move correctly
lxt move to 2ns
```

**Expected Behavior**

lxt_ttc should always move each motor to the correct position

**Context**

Makes it impossible to use these in timing scans, despite having worked previously!

### LINKED ISSUE #772: lxt_ttc.set_current_position (XPP, XCS)
**Feature Request**

The lxt_ttc motor in XPP/XCS should have a `set_current_position` method that correctly sets the current position of the component lxt and txt motors.

**Existing Alternatives**

Need to call `set_current_position` independently on each of the component motors.

**Context**

This could probably be handled generically for pseudopositioners

### LINKED ISSUE #776: Add warning for inconsistent lxt_ttc
**Feature Request**

When lxt_ttc is in an inconsistent state prior to a move (delay and timing are not equal/opposite), show a warning and ask if you want Python to move txt to fix this.

**Existing Alternatives**

Currently it just "fixes" itself immediately, which may not be what you want.

**Context**

Request from Diling

### LINKED ISSUE #777: lxt_ttc should be defined specially per instrument
**Feature Request**

Currently, the lxt_ttc class is defined very specifically in pcdsdevices (newport delay with 2 bounces and vitara timing)
But this class isn't very clear and shouldn't even be included in pcdsdevices, rather each hutch should have something like this in their beamline.py file:
```
class MyDelayClass(DelayBase):
    motor = Cpt(IMS, '')

class TTC(SyncAxesBase):
    txt = Cpt(LaserTiming, MY:XPP:PV, <some kwargs>)
    lxt = Cpt(MyDelayClass, MY:XPP:PV2, n_bounces=14, invert=True, <some kwargs>)

lxt_ttc = TTC('', name='lxt_ttc')
```
This allows the hutch to have better agency over how to define the scripting for their endstation or experiment-specific delay scanning

**Existing Alternatives**

Modify random properties on `pcdsdevices.lxe.LaserTimingCompensation` post-instantiation

**Context**

Based on conversation with Diling

### LINKED ISSUE #778: Add an invert/direction parameter to DelayBase
**Feature Request**

Allow us to define whether increasing the motor value increases or decreases the delay

**Existing Alternatives**

Could reverse the underlying EPICS motor?
Could make special subclasses? https://github.com/pcdshub/pcdsdevices/blob/3b3041da26bcf89ae050ead6c9d06bda6b3e6b8a/pcdsdevices/lxe.py#L376

**Context**

Requested by Diling in zoom

### PR DESCRIPTION
**Pull request by ZLLentz**
## Description

This PR handles many of the requests for `lxt_ttc` by expanding the sync motor handling. To go through them one by one:

- Fix issues where `lxt_ttc` could have inconsistent offset behavior by allowing the `SyncAxis` to define static offsets (which is going to be the default). Since I'm changing a default, I've decided to deprecate the previous `SyncAxesBase`. (closes #770)
- Add `set_current_position` to all pcdsdevices `PseudoPositioner`s (closes #772)
- Add a warning when any `SyncAxis` subclass has motor positions that do not match each other properly, and provide a method (`fix_sync`) to re-synchronize them (closes #776) 
- Provide an example of a clean way to implement `lxt_ttc` per instrument so that the scientists can reconfigure it as needed (closes #777)
- Add direction inversion directly to `DelayBase` instead of requiring a special inversion class. (closes #778)
- Add `set_position` as an alias to `set_current_position` (no issue number, but I remember this from the meeting)

Obsoletes the following PRs:
- closes #771
- closes #773

## Motivation and Context

`lxt_ttc` had issues recently that made the code unusable, we want to:
- fix the issues
- use a more stable and configurable base
- allow the scientists to configure `lxt_ttc` themselves, as it will change between experiments
- add requested features

## How Has This Been Tested?

Added some unit tests

## Where Has This Been Documented?

Added pre-release docs

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on travis
- [x] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

**Comment by ZLLentz** [comment]
This is more or less where I want to be with this PR, it has all the things needed for the `lxt_ttc` stuff and more, but generalized into `SyncAxis`. I'm going to add tests and maybe some more docs before I call it done.

I'd like feedback on:
- Was this even a good idea at all?
- Implementation quality/issues

File: pcdsdevices/pseudopos.py [hunk 1]
```diff
@@ -282,6 +309,314 @@ def inverse(self, real_pos):
...
+                raise ValueError(
+                    'Invalid fix_sync_keep_still. Must match a motor.'
+                    ) from None
+
+    def _check_info_dict(self, setting, info_kind):
+        """Helper function to check that all keys in the dict are Cpts"""
+        if setting is not None:
+            try:
+                for attr, _ in setting.items():
+                    getattr(self.__class__, attr)
```
Comment (klauer) [inline pcdsdevices/pseudopos.py:469]: Perhaps `if attr not in self.component_names:` instead?

File: pcdsdevices/pseudopos.py [hunk 2]
```diff
@@ -282,6 +309,314 @@ def inverse(self, real_pos):
...
+                self.offset_mode = SyncAxisOffsetMode[self.offset_mode]
+            except KeyError:
+                raise ValueError(
+                    f'Invalid offset_mode: {self.offset_mode}'
+                    ) from None
+        self._check_info_dict(self.offsets, 'offsets')
+        self._check_info_dict(self.scales, 'scales')
+        if self.fix_sync_keep_still is not None:
+            try:
+                getattr(self.__class__, self.fix_sync_keep_still)
```
Comment (klauer) [inline pcdsdevices/pseudopos.py:458]: Perhaps `self.fix_sync_keep_still in self.component_names`?

It's set at class definition time, so it should be available here.

File: pcdsdevices/pseudopos.py [hunk 3]
```diff
@@ -282,6 +309,314 @@ def inverse(self, real_pos):
...
+        """Helper function to fill default values into the dict"""
+        if setting is None:
+            setting = {}
+        elif isinstance(setting, dict):
+            setting = copy.copy(setting)
+        else:
+            raise ValueError(
+                f'Invalid {info_kind}: {setting}, must be dict or None')
+        for attr in self._real_attrs:
+            if attr not in setting:
```
Comment (klauer) [inline pcdsdevices/pseudopos.py:485]: `setting.setdefault(attr, default)` is kind of nice shorthand for this:

```
In [2]: a = {}

In [3]: a.setdefault('a', 4)
Out[3]: 4

In [4]: a
Out[4]: {'a': 4}

In [5]: a['b'] = 5

In [6]: a.setdefault('b', 7)
Out[6]: 5

In [7]: a
Out[7]: {'a': 4, 'b': 5}
```

File: pcdsdevices/pseudopos.py [hunk 4]
```diff
@@ -282,6 +309,314 @@ def inverse(self, real_pos):
...
+        This gives us the sync readback position.
+        """
+        return (pos - self.offsets[attr]) / self.scales[attr]
+
+    def consistency_warning(self):
+        """
+        Let the user know that the axes are desync'd
+        """
+        if self.warn_inconsistent:
+            logger.warning(
```
Comment (klauer) [inline pcdsdevices/pseudopos.py]:
```suggestion
            self.log.warning(
```

**Review by klauer — COMMENTED** [review]
I really like this implementation. It's got a lot of flexibility and it's easily configurable.

Some optional style/code suggestions, but none of it has any bearing on the actual implementation.

**Comment by klauer** [comment]
And just noticed a commit came in while I was reviewing - including that info in the status repr makes a lot of sense. 👍

File: pcdsdevices/pseudopos.py (hunk 4 above)
Comment (ZLLentz) [reply pcdsdevices/pseudopos.py]: I always forget about the self.log

File: pcdsdevices/pseudopos.py (hunk 2 above)
Comment (ZLLentz) [reply pcdsdevices/pseudopos.py:458]: I'll do this, avoiding the dunder variables is preferred when possible

File: pcdsdevices/pseudopos.py (hunk 1 above)
Comment (ZLLentz) [reply pcdsdevices/pseudopos.py:469]: I'll also do this

File: pcdsdevices/pseudopos.py (hunk 3 above)
Comment (ZLLentz) [reply pcdsdevices/pseudopos.py:485]: Never knew about this, will revise accordingly

**Comment by ZLLentz** [comment]
That's the last thing I wanted to do here, should probably see if we can take `LxtTtcExample` for a test drive in XPP and copy that definition to XPP's beamline file

**Comment by ZryletTC** [comment]
Yes, I'll try to test in XPP later today. They might want to run woth the old way for this experiment though.

**Comment by ZLLentz** [comment]
Cool, whatever they want to do is fine

**Comment by ZLLentz** [comment]
This should be done

**Comment by ZLLentz** [comment]
If there are no objections, I'll merge before I stop for the day

**Comment by klauer** [comment]
I take it @ZryletTC 's tests were successful? If so, I have nothing further to add here.

**Review by ZryletTC — APPROVED** [review]
They most basic tests are good. It doesn't break everything and an object can be created without errors. Any further implementation stuff can be handled in a new PR if necessary.

**Review by klauer — APPROVED** [review]

**Comment by ZLLentz** [comment]
Resolving merge conflict and confirming no CI mishap before merging

**Pull request merged**